concordium-consensus/src/Concordium/KonsensusV1/Consensus/Timeout.hs

```diff
@@ -0,0 +1,534 @@
+    finIndex = tmFinalizerIndex $ tmBody tm
+    singletonTimeout = Map.singleton finIndex tm
+    insertTimeout = Map.insert finIndex tm
+
+-- |Process a timeout message. This stores the timeout, and makes sure the stored timeout messages
+-- do not span more than 2 epochs. If enough timeout messages are stored, we form a timeout certificate and
+-- advance round.
+--
+-- Precondition:
+-- * The given 'TimeoutMessage' is valid and has already been checked.
+processTimeout ::
+    ( MonadTimeout m,
+      LowLevel.MonadTreeStateStore m,
+      MonadState (SkovData (MPV m)) m
+    ) =>
+    TimeoutMessage ->
+    m ()
+processTimeout tm = do
+    timeoutMessages <- use currentTimeoutMessages
+    currentRoundStatus <- use roundStatus
+    let highestCB = _rsHighestCertifiedBlock currentRoundStatus
+    -- Add the new timeout message to the current messages.
+    -- If the result is 'Nothing', then there was no change as a result, so nothing left to do.
+    let maybeNewTimeoutMessages = updateTimeoutMessages timeoutMessages tm
+    forM_ maybeNewTimeoutMessages $ \newTimeoutMessages@TimeoutMessages{..} -> do
+        currentTimeoutMessages .=! Present newTimeoutMessages
+        skovData <- get
+        let getFinalizersForEpoch epoch = (^. bfFinalizers) <$> getBakersForEpoch epoch skovData
+        -- We should not fail to get the finalizers for the epoch of the highest QC, because it
+        -- should either be the current epoch or the previous one.
+        let maybeFinComm = getFinalizersForEpoch (cbEpoch highestCB)
+        forM_ maybeFinComm $ \finCommQC -> do
+            -- The baker IDs of the finalizers who have signed in the first epoch.
+            let firstBakerIds
+                    | Just firstFinComm <- getFinalizersForEpoch tmFirstEpoch =
+                        bakerIdsFor firstFinComm tmFirstEpochTimeouts
+                    | otherwise = Set.empty
+            -- The baker IDs of the finalizers who have signed in the second epoch.
+            let secondBakerIds
+                    | not (null tmSecondEpochTimeouts),
+                      Just secondFinComm <- getFinalizersForEpoch (tmFirstEpoch + 1) =
+                        bakerIdsFor secondFinComm tmSecondEpochTimeouts
+                    | otherwise = Set.empty
+            -- Compute the voter power in the epoch of the highest QC for a baker by the baker ID.
+            let getBakerVoterPower = fmap finalizerWeight . finalizerByBakerId finCommQC
+            let voterPowerSum =
+                    sum . mapMaybe getBakerVoterPower $
+                        Set.toList (firstBakerIds `Set.union` secondBakerIds)
+            let totalWeightRational = toRational $ committeeTotalWeight finCommQC
+            genesisSigThreshold <- toRational . genesisSignatureThreshold . gmParameters <$> use genesisMetadata
+            let voterPowerSumRational = toRational voterPowerSum
+            when (voterPowerSumRational / totalWeightRational >= genesisSigThreshold) $ do
```

Comment:
Great observation. As you described, there's some complexity to this as the weights could be computed differently. 

I don't think it should be a priority right now.

I've added this to the list of optimizations that can be done later: https://github.com/Concordium/concordium-node/issues/781

Reply:
When is "later"? It doesn't have to be done this week, but I don't think that it makes sense to start benchmarking the protocol without this change.

The original Jolteon protocol that we found in the literature had complexity `O(n^2 log n)`, and we went to some trouble to change parts to bring it down to `O(n log n)`. You went to a lot of trouble to store which QCs have been checked, so that we don't recheck them and bring the complexity from `O(n^2 log n)` to `O(n log n)`. All that to have it blow up to `O(n^2 log n)` again, because we recompute the weights the whole time instead of storing the running total seems a little silly.

For 40 finalizers it might not make much difference. But one of the things we need to do is run simulations with 500 or 1000 finalizers, and for that there could be an impact.

Reply:
We can of course benchmark now and then the fixed version to see how much it improves. That would actually also be interesting. ;)

Reply:
We will have to figure that out, i.e. **when**. 

However I think it should at least be postponed until we got a working version (and hence this is merged in together with the two other PRs (scheduler and integration into the node)